### PR TITLE
Allow truststore.SSLContext to be used globally

### PR DESCRIPTION
**Pull request by sethmlarson**
Closes https://github.com/sethmlarson/truststore/issues/75

cc @guerda @pfeigl @JonasR

### CODE REVIEW

**Comment by davisagli** [comment]
I haven't done a close review yet, but I think this is moving in the direction of what most users of truststore probably want (secure by default, without needing to pay attention to specific call sites) and I'm happy to see it in progress.

I do have some concern because of the fact that the patching is dependent on import order (i.e. if a module imports SSLContext into its own namespace before the patch is injected, it won't take effect there). There are some devious ways to try to solve that which I wouldn't necessarily recommend (i.e. following references via `gc.get_referrers` or replacing the builtin SSLContext's `__new__` method). Another possibility that occurred to me would be to replace the openssl verify callback using ctypes the way that Steve Dower showed in https://gist.github.com/zooba/bce89eaa865bf89611f7dab9a2a38276, instead of replacing all of SSLContext. But that would be a major refactoring and I haven't checked whether it would be compatible with what's needed for the Mac and Windows implementations.

**Comment by sethmlarson** [comment]
@davisagli I'm interested in trying out swapping the verify callback, I'm assuming that would be essentially the next step for this library if we're planning on making it into the stdlib. Let's create a separate issue and investigate? I'd hope we'd be able to access the full chain at that point in the call, if so we can probably do what we're doing currently on macOS and Windows.

**Comment by sethmlarson** [comment]
Okay this is ready for a review @davisagli. I'm going to open an issue to track down why using truststore.SSLContext as a server doesn't work on specifically macOS. I don't have one of those handy and testing via CI was getting tedious.

File: docs/source/index.md [hunk 1]
```diff
@@ -20,88 +30,143 @@ This project should be considered experimental so shouldn't be used in productio
...
 ```{code-block} python
-   import ssl
-   import aiohttp
-   import truststore
+import aiohttp
+import truststore
+
+truststore.inject_into_ssl()
+
+http = aiohttp.ClientSession(ssl=ctx)
```
Comment (sethmlarson) [inline docs/source/index.md:109]: Need to remove ctx

File: src/truststore/_api.py [hunk 2]
```diff
@@ -13,16 +19,54 @@
...
+
 
 class SSLContext(ssl.SSLContext):
     """SSLContext API that uses system certificates on all platforms"""
 
-    def __init__(self, protocol: int = ssl.PROTOCOL_TLS) -> None:
-        self._ctx = ssl.SSLContext(protocol)
+    def __init__(self, protocol: int = None) -> None:  # type: ignore[assignment]
+        self._ctx = _original_SSLContext(protocol)
+        _configure_context(self._ctx)
```
Comment (sethmlarson) [inline src/truststore/_api.py:69]: Remove this

File: README.md [hunk 3]
```diff
@@ -3,49 +3,60 @@
...
+- Linux via OpenSSL
 
-# Also works with libraries that accept an SSLContext object
-import urllib3
+## User Guide
 
-http = urllib3.PoolManager(ssl_context=ctx)
-http.request("GET", "https://example.com")
+You can inject `truststore` into the standard library `ssl` module so the functionality is used
+by every library by default. To do so use the `truststore.inject_into_ssl()` function:
```
Comment (davisagli) [inline README.md:26]: @sethmlarson We should mention the importance of doing this as early as possible / the caveat that the injection won't impact modules that already imported SSLContext.

File: docs/source/index.md [hunk 4]
```diff
@@ -5,103 +5,171 @@
...
+```{code-block} python
+import requests
+import truststore
+
+truststore.inject_into_ssl()
+
+resp = requests.request("GET", "https://example.com")
+```
+
 Requests doesn't support passing an {py:class}`ssl.SSLContext` object to a `requests.Session` object directly so there's an additional class you need to inject the `truststore.SSLContext` instance to the lower-level {py:class}`urllib3.PoolManager` instance:
```
Comment (davisagli) [inline docs/source/index.md:143]: Is this still needed if you did the injection above, or only if you didn't? It's not clear.

**Review by davisagli — COMMENTED** [review]
The implementation LGTM. Just a few suggestions to clarify the docs.

File: docs/source/index.md (hunk 4 above)
Comment (sethmlarson) [reply docs/source/index.md:143]: Yeah this section is just a big hack, going to remove the whole thing. Users of requests should use global injection and be fine.

File: README.md (hunk 3 above)
Comment (sethmlarson) [reply README.md:26]: That's a good idea, thank you!

**Pull request merged**